internal/error.go

```diff
@@ -0,0 +1,80 @@
+package fleetlock
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+)
+
+// ErrorKind is a error type identifier.
+//
+// It is designed to be tracked in Zincati metrics.
+// Fleetlock specification states that this MUST have a bounded/small cardinality.
+type ErrorKind string
+
+// List of ErrorKinds
+const (
+	ErrorMethodNotAllowed ErrorKind = "method_not_allowed"
+	ErrorMissingHeader    ErrorKind = "missing_header"
+	ErrorDecodingRequest  ErrorKind = "failed_decoding_request"
+	ErrorInternal         ErrorKind = "internal_error"
+	ErrorLocked           ErrorKind = "locked"
+)
+
+// Error holds data used for http responses in case of error.
+type Error struct {
+	// error type identifier
+	Kind ErrorKind `json:"kind"`
+
+	// human-friendly error description
+	Message string `json:"value"`
+
+	StatusCode int `json:"-"`
+}
+
+// NewError creates an Error with a error kind, and a templated error message.
+func NewError(kind ErrorKind, format string, a ...interface{}) Error {
+	var statusCode int
+	switch kind {
+	case ErrorMethodNotAllowed:
+		statusCode = http.StatusMethodNotAllowed
+	case ErrorDecodingRequest, ErrorMissingHeader:
+		statusCode = http.StatusBadRequest
+	case ErrorInternal:
+		statusCode = http.StatusInternalServerError
+	case ErrorLocked:
+		statusCode = http.StatusLocked
+	default:
+		statusCode = http.StatusInternalServerError
+	}
+
+	message := fmt.Sprintf(format, a...)
+	return Error{
+		Kind:       kind,
+		Message:    message,
+		StatusCode: statusCode,
+	}
+}
+
+// Print is used to print the error HTTP response (JSON output).
+func (e Error) Print(w http.ResponseWriter) error {
```

Comment:
Move this stuff into server.Error, the message type doesn't need to be concerned with HTTP